src/connections/allContacts.js

```diff
@@ -1,5 +1,62 @@
 import { connect, withMutations } from 'cozy-client'
 import flow from 'lodash/flow'
+import array from 'lodash/array'
+import isEqual from 'lodash/isEqual'
+
+function flatten(arrayOfArrays) {
+  return [].concat(...arrayOfArrays)
+}
+
+function removeDuplicates(data) {
+  return array.uniqWith(data, isEqual)
+}
+
+function findItems(doctype, selectors) {
+  return client =>
+    Promise.all(
+      selectors.map(selector =>
+        client.query(client.find(doctype).where(selector))
+      )
+    )
+      .then(results => results.map(result => result.data))
+      .then(flatten)
+      .then(removeDuplicates)
+}
+
+function selector(attr, prop) {
+  return values => ({
+    [attr]: {
+      $elemMatch: {
+        [prop]: {
+          $in: values
+        }
+      }
+    }
+  })
+}
+
+function getValues(array = [], property) {
+  return array.map(getValue(property))
+}
+
+function getValue(property) {
+  return obj => obj[property]
+}
+
+function createSelector(object) {
+  return path => {
+    const [arrayName, nestedProperty] = path.split('.')
+    const arrayOfObject = object[arrayName]
+    const values = getValues(arrayOfObject, nestedProperty)
+    return selector(arrayName, nestedProperty)(values)
+  }
+}
```

Comment:
Every functions above could be extracted to be part of an external library as told here https://github.com/cozy/cozy-contacts/pull/241#discussion_r212284135

Reply:
They make the code easier to read and understand by offering only small functions that do not do much, and they allow testing the different steps of the process by simple unit tests.

Reply:
I missed the earlier implementation, but this one looks pretty good to me. However I would agree with @ptbrowne that the currying makes things really hard to understand. Making a mental model of what `const selectors = paths.map(createSelector(targetContact))` does took me a while, and it looks like I'm not alone.

Maybe we could simply un-curry the function, ie. `selectFromObject(object, path)`? And I guess `getValues` could do the mapping directly rather than relying on the singular `getValue`.

If it stays like this, I think I'd feel a lot more confident with some unit tests, in case we change the implementation later on.

Does that sound ok?

Reply:
Je vais tenter de défendre un de mes choix, mais si je lis l'instruction, je vois qu'on crée une variable `selectors` qui va probablement contenir des `selectors` qui sont construits de la sorte : pour chaque élément contenu dans `paths`, on crée un `selector` avec la fonction `createSelector` et `targetContact`.
C'est du sucre syntaxique le currying mais je veux bien croire que ça puisse être déroutant au premier abord, comme l'étaient les arrow functions ou les opérateurs `...`.